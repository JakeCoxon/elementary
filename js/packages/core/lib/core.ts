import {
  createNode,
  isNode,
  resolve,
} from '../nodeUtils';

import type {NodeRepr_t} from '../src/Reconciler.gen';


// Generic
type OptionalKeyProps = {
  key?: string,
};

// Const node and constant value nodes
type ConstNodeProps = {
  key?: string,
  value: number,
};

export function constant(props: ConstNodeProps): NodeRepr_t {
  return createNode("const", props, []);
}

export function sr(): NodeRepr_t {
  return createNode("sr", {}, []);
}

export function time(): NodeRepr_t {
  return createNode("time", {}, []);
}

// Counter node
export function counter(gate: NodeRepr_t | number): NodeRepr_t;
export function counter(props: OptionalKeyProps, gate: NodeRepr_t | number): NodeRepr_t;
export function counter(a, b?) {
  if (typeof a === "number" || isNode(a)) {
    return createNode("counter", {}, [resolve(a)]);
  }

  return createNode("counter", a, [resolve(b)]);
}

// Accum node
export function accum(xn: NodeRepr_t | number, reset: NodeRepr_t | number): NodeRepr_t;
export function accum(props: OptionalKeyProps, xn: NodeRepr_t | number, reset: NodeRepr_t | number): NodeRepr_t;
export function accum(a, b, c?) {
  if (typeof a === "number" || isNode(a)) {
    return createNode("accum", {}, [resolve(a), resolve(b)]);
  }

  return createNode("accum", a, [resolve(b), resolve(c)]);
}

// Phasor node
export function phasor(rate: NodeRepr_t | number, reset: NodeRepr_t | number): NodeRepr_t;
export function phasor(props: OptionalKeyProps, rate: NodeRepr_t | number, reset: NodeRepr_t | number): NodeRepr_t;
export function phasor(a, b, c?) {
  if (typeof a === "number" || isNode(a)) {
    return createNode("phasor", {}, [resolve(a), resolve(b)]);
  }

  return createNode("phasor", a, [resolve(b), resolve(c)]);
}

// Latch node
export function latch(t: NodeRepr_t | number, x: NodeRepr_t | number): NodeRepr_t;
export function latch(props: OptionalKeyProps, t: NodeRepr_t | number, x: NodeRepr_t | number): NodeRepr_t;
export function latch(a, b, c?) {
  if (typeof a === "number" || isNode(a)) {
    return createNode("latch", {}, [resolve(a), resolve(b)]);
  }

  return createNode("latch", a, [resolve(b), resolve(c)]);
}

// MaxHold node
type MaxHoldNodeProps = {
  key?: string,
  hold?: number,
};

export function maxhold(x: NodeRepr_t | number, reset: NodeRepr_t | number): NodeRepr_t;
export function maxhold(props: MaxHoldNodeProps, x: NodeRepr_t | number, reset: NodeRepr_t | number): NodeRepr_t;
export function maxhold(a, b, c?) {
  if (typeof a === "number" || isNode(a)) {
    return createNode("maxhold", {}, [resolve(a), resolve(b)]);
  }

  return createNode("maxhold", a, [resolve(b), resolve(c)]);
}

// Once node
type OnceNodeProps = {
  key?: string,
  arm?: boolean,
};

export function once(x: NodeRepr_t | number): NodeRepr_t;
export function once(props: OnceNodeProps, x: NodeRepr_t | number): NodeRepr_t;
export function once(a, b?) {
  if (typeof a === "number" || isNode(a)) {
    return createNode("once", {}, [resolve(a)]);
  }

  return createNode("once", a, [resolve(b)]);
}

// Rand node
type RandNodeProps = {
  key?: string,
  seed?: number,
};

export function rand(): NodeRepr_t;
export function rand(props: RandNodeProps): NodeRepr_t;
export function rand(a?) {
  if (typeof a !== "undefined") {
    return createNode("rand", a, []);
  }

  return createNode("rand", {}, []);
}

// Metro node
type MetroNodeProps = {
  key?: string,
  interval?: number,
};

export function metro(): NodeRepr_t;
export function metro(props: MetroNodeProps): NodeRepr_t;
export function metro(a?) {
  if (typeof a !== "undefined") {
    return createNode("metro", a, []);
  }

  return createNode("metro", {}, []);
}

// Sample node
type SampleNodeProps = {
  key?: string,
  path?: string,
  mode?: string,
  startOffset?: number,
  stopOffset?: number,
};

export function sample(props: SampleNodeProps, trigger: NodeRepr_t | number, rate: NodeRepr_t | number): NodeRepr_t {
  return createNode("sample", props, [resolve(trigger), resolve(rate)]);
}

// Table node
type TableNodeProps = {
  key?: string,
  path?: string,
};

export function table(props: TableNodeProps, t: NodeRepr_t | number): NodeRepr_t {
  return createNode("table", props, [resolve(t)]);
}

// Convolve node
type ConvolveNodeProps = {
  key?: string,
  path?: string,
};

export function convolve(props: ConvolveNodeProps, x: NodeRepr_t | number): NodeRepr_t {
  return createNode("convolve", props, [resolve(x)]);
}

// Seq node
type SeqNodeProps = {
  key?: string,
  seq?: Array<number>,
  offset?: number,
  hold?: boolean,
  loop?: boolean,
};

export function seq(props: SeqNodeProps, trigger: NodeRepr_t | number, reset: NodeRepr_t | number): NodeRepr_t {
  return createNode("seq", props, [resolve(trigger), resolve(reset)]);
}

// Seq2 node
type Seq2NodeProps = {
  key?: string,
  seq?: Array<number>,
  offset?: number,
  hold?: boolean,
  loop?: boolean,
};

export function seq2(props: Seq2NodeProps, trigger: NodeRepr_t | number, reset: NodeRepr_t | number): NodeRepr_t {
  return createNode("seq2", props, [resolve(trigger), resolve(reset)]);
}

// SparSeq node
type SparSeqNodeProps = {
  key?: string,
  seq?: Array<{value: number, tickTime: number}>,
  offset?: number,
  loop?: boolean | Array<number>,
  resetOnLoop?: boolean,
  interpolate?: number,
  tickInterval?: number,
};

export function sparseq(props: SparSeqNodeProps, trigger: NodeRepr_t | number, reset: NodeRepr_t | number): NodeRepr_t {
  return createNode("sparseq", props, [resolve(trigger), resolve(reset)]);
}

// Pole node
export function pole(p: NodeRepr_t | number, x: NodeRepr_t | number): NodeRepr_t;
export function pole(props: OptionalKeyProps, p: NodeRepr_t | number, x: NodeRepr_t | number): NodeRepr_t;
export function pole(a, b, c?) {
  if (typeof a === "number" || isNode(a)) {
    return createNode("pole", {}, [resolve(a), resolve(b)]);
  }

  return createNode("pole", a, [resolve(b), resolve(c)]);
}

// Env node
export function env(atkPole: NodeRepr_t | number, relPole: NodeRepr_t | number, x: NodeRepr_t | number): NodeRepr_t;
export function env(props: OptionalKeyProps, atkPole: NodeRepr_t | number, relPole: NodeRepr_t | number, x: NodeRepr_t | number): NodeRepr_t;
export function env(a, b, c, d?) {
  if (typeof a === "number" || isNode(a)) {
    return createNode("env", {}, [resolve(a), resolve(b), resolve(c)]);
  }

  return createNode("env", a, [resolve(b), resolve(c), resolve(d)]);
}

// Single sample delay node
export function z(x: NodeRepr_t | number): NodeRepr_t;
export function z(props: OptionalKeyProps, x: NodeRepr_t | number): NodeRepr_t;
export function z(a, b?) {
  if (typeof a === "number" || isNode(a)) {
    return createNode("z", {}, [resolve(a)]);
  }

  return createNode("z", a, [resolve(b)]);
}

// Variable delay node
type DelayNodeProps = {
  key?: string,
  size: number,
};

export function delay(
  props: DelayNodeProps,
  len: NodeRepr_t | number,
  fb: NodeRepr_t | number,
  x : NodeRepr_t | number,
): NodeRepr_t;

export function delay(a, b, c, d) {
  if (typeof a === "number" || isNode(a)) {
    return createNode("delay", {}, [resolve(a), resolve(b), resolve(c)]);
  }

  return createNode("delay", a, [resolve(b), resolve(c), resolve(d)]);
}

export function sdelay(props: DelayNodeProps, x: NodeRepr_t | number): NodeRepr_t {
  return createNode("sdelay", props, [resolve(x)]);
}

// SVF
export function svf(
  fc: NodeRepr_t | number,
  q: NodeRepr_t | number,
  x: NodeRepr_t | number,
): NodeRepr_t;

export function svf(
  props: {
    key?: string,
    mode?: string,
  },
  fc: NodeRepr_t | number,
  q: NodeRepr_t | number,
  x: NodeRepr_t | number,
): NodeRepr_t;

export function svf(a, b, c, d?) {
  if (typeof a === "number" || isNode(a)) {
    return createNode("svf", {}, [
      resolve(a),
      resolve(b),
      resolve(c),
    ]);
  }

  return createNode("svf", a, [
    resolve(b),
    resolve(c),
    resolve(d),
  ]);
}

export function svfshelf(
  fc: NodeRepr_t | number,
  q: NodeRepr_t | number,
  gainDecibels: NodeRepr_t | number,
  x: NodeRepr_t | number,
): NodeRepr_t;

export function svfshelf(
  props: {
    key?: string,
    mode?: string,
  },
  fc: NodeRepr_t | number,
  q: NodeRepr_t | number,
  gainDecibels: NodeRepr_t | number,
  x: NodeRepr_t | number,
): NodeRepr_t;

export function svfshelf(a, b, c, d, e?) {
  if (typeof a === "number" || isNode(a)) {
    return createNode("svfshelf", {}, [
      resolve(a),
      resolve(b),
      resolve(c),
      resolve(d),
    ]);
  }

  return createNode("svfshelf", a, [
    resolve(b),
    resolve(c),
    resolve(d),
    resolve(e),
  ]);
}

// Biquad node
export function biquad(
  b0: NodeRepr_t | number,
  b1: NodeRepr_t | number,
  b2: NodeRepr_t | number,
  a1: NodeRepr_t | number,
  a2: NodeRepr_t | number,
  x: NodeRepr_t | number,
): NodeRepr_t;

export function biquad(
  props: OptionalKeyProps,
  b0: NodeRepr_t | number,
  b1: NodeRepr_t | number,
  b2: NodeRepr_t | number,
  a1: NodeRepr_t | number,
  a2: NodeRepr_t | number,
  x: NodeRepr_t | number,
): NodeRepr_t;

export function biquad(a, b, c, d, e, f, g?) {
  if (typeof a === "number" || isNode(a)) {
    return createNode("biquad", {}, [
      resolve(a),
      resolve(b),
      resolve(c),
      resolve(d),
      resolve(e),
      resolve(f),
    ]);
  }

  return createNode("biquad", a, [
    resolve(b),
    resolve(c),
    resolve(d),
    resolve(e),
    resolve(f),
    resolve(g),
  ]);
}

// Feedback nodes
type TapInNodeProps = {
  name: string,
};

type TapOutNodeProps = {
  name: string,
};

export function tapIn(props: TapInNodeProps): NodeRepr_t {
  return createNode("tapIn", props, []);
}

export function tapOut(props: TapOutNodeProps, x: NodeRepr_t | number): NodeRepr_t {
  return createNode("tapOut", props, [resolve(x)]);
}

// Meter node
type MeterNodeProps = {
  key?: string,
  name?: string,
};

export function meter(x: NodeRepr_t | number): NodeRepr_t;
export function meter(props: MeterNodeProps, x: NodeRepr_t | number): NodeRepr_t;
export function meter(a, b?) {
  if (typeof a === "number" || isNode(a)) {
    return createNode("meter", {}, [resolve(a)]);
  }

  return createNode("meter", a, [resolve(b)]);
}

// Snapshot node
type SnapshotNodeProps = {
  key?: string,
  name?: string,
};

export function snapshot(trigger: NodeRepr_t | number, x: NodeRepr_t | number): NodeRepr_t;
export function snapshot(props: SnapshotNodeProps, trigger: NodeRepr_t | number, x: NodeRepr_t | number): NodeRepr_t;
export function snapshot(a, b, c?) {
  if (typeof a === "number" || isNode(a)) {
    return createNode("snapshot", {}, [resolve(a), resolve(b)]);
  }

  return createNode("snapshot", a, [resolve(b), resolve(c)]);
}

// Scope node
type ScopeNodeProps = {
  key?: string,
  name?: string,
  size?: number,
  channels?: number,
};

export function scope(...args : Array<NodeRepr_t | number>): NodeRepr_t;
export function scope(props: ScopeNodeProps, ...args : Array<NodeRepr_t | number>): NodeRepr_t;
export function scope(a, ...bs) {
  if (typeof a === "number" || isNode(a)) {
    return createNode("scope", {}, [a, ...bs].map(resolve));
  }

  return createNode("scope", a, bs.map(resolve));
}

// FFT node
type FFTNodeProps = {
  key?: string,
  name?: string,
  size?: number,
};

export function fft(x: NodeRepr_t | number): NodeRepr_t;
export function fft(props: FFTNodeProps, x: NodeRepr_t | number): NodeRepr_t;
export function fft(a, b?) {
  if (typeof a === "number" || isNode(a)) {
    return createNode("fft", {}, [resolve(a)]);
  }

  return createNode("fft", a, [resolve(b)]);
}
