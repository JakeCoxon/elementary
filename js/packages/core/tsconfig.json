{
  "compilerOptions": {
    "allowJs": true,
    "declaration": true,
    "target": "es6",
    "module": "es6",
    "moduleResolution": "node",
  },
  "files": [
    "index.ts",
  ]
}
