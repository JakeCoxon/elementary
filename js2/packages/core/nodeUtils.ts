import {
  NodeRepr_createPrimitive,
  NodeRepr_createComposite,
  NodeRepr_isNode,
} from './src/Reconciler.gen';

import invariant from 'invariant';

import type {NodeRepr_t} from './src/Reconciler.gen';


export function resolve(n : NodeRepr_t | number): NodeRepr_t {
  return typeof n === 'number'
    ? NodeRepr_createPrimitive("const", {value: n}, [])
    : n;
}

export function isNode(n: any): boolean {
  return NodeRepr_isNode(n);
}

export function createNode(
  kind: Parameters<typeof NodeRepr_createPrimitive>[0] | Parameters<typeof NodeRepr_createComposite>[0],
  props,
  children: Array<NodeRepr_t | number>
): NodeRepr_t {
  invariant(children.length <= 8, `Nodes can only have at most 8 children.`);

  if (typeof kind === 'string') {
    return NodeRepr_createPrimitive(kind, props, children.map(resolve));
  }

  return NodeRepr_createComposite(kind, props, children.map(resolve));
}
