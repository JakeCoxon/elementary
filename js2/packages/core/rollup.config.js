import cjs from '@rollup/plugin-commonjs';
import dts from 'rollup-plugin-dts';
import { nodeResolve } from '@rollup/plugin-node-resolve';
import { terser } from 'rollup-plugin-terser';

export default [
  // Bundle code
  {
    input: 'dist/index.js',
    output: {
      file: './dist/main.min.js',
    },
    plugins: [
      cjs(),
      nodeResolve(),
      terser(),
    ],
  },
  // Bundle types
  {
    input: 'dist/index.d.ts',
    output: {
      file: './dist/main.min.d.ts',
    },
    plugins: [
      dts(),
    ],
  },
]
