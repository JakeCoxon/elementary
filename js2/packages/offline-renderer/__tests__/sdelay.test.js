import OfflineRenderer from '../index';
import { el } from '@elemaudio/core';


test('sdelay basics', async function() {
  let core = new OfflineRenderer();

  await core.initialize({
    numInputChannels: 1,
    numOutputChannels: 1,
  });

  // Graph
  core.render(el.sdelay({size: 10}, el.in({channel: 0}), 0));

  // Ten blocks of data
  let inps = [new Float32Array(512 * 10)];
  let outs = [new Float32Array(512 * 10)];

  // Get past the fade-in
  core.process(inps, outs);

  // Now we use a very calculated input to step the sparseq
  inps = [Float32Array.from([
    1, 2, 3, 4, 4, 3, 2, 1,
    0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0,
  ])];

  outs = [new Float32Array(inps[0].length)];

  // Drive our sdelay
  core.process(inps, outs);

  expect(outs[0]).toMatchSnapshot();
});
