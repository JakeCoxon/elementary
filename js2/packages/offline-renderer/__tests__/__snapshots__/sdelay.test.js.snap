// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sdelay basics 1`] = `
Float32Array [
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  1,
  2,
  3,
  4,
  4,
  3,
  2,
  1,
  0,
  0,
  0,
  0,
  0,
  0,
]
`;
